hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/upgrade/FinalizeScmUpgradeSubcommand.java

```diff
@@ -0,0 +1,153 @@
+import static org.apache.hadoop.hdds.scm.cli.upgrade.FinalizeUpgradeCommandUtil.isStarting;
+
+import java.io.IOException;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import org.apache.hadoop.hdds.cli.HddsVersionProvider;
+import org.apache.hadoop.hdds.scm.cli.ScmSubcommand;
+import org.apache.hadoop.hdds.scm.client.ScmClient;
+import org.apache.hadoop.ozone.upgrade.UpgradeException;
+import org.apache.hadoop.ozone.upgrade.UpgradeFinalizer;
+import org.apache.hadoop.ozone.upgrade.UpgradeFinalizer.StatusAndMessages;
+
+import picocli.CommandLine;
+
+/**
+ * Handler of Finalize SCM command.
+ */
+@CommandLine.Command(
+    name = "finalizeupgrade",
+    description = "Finalize SCM Upgrade",
+    mixinStandardHelpOptions = true,
+    versionProvider = HddsVersionProvider.class)
+
+public class FinalizeScmUpgradeSubcommand extends ScmSubcommand {
+  @CommandLine.Option(
+      names = {"--takeover"},
+      description = "Forces takeover of monitoring from another client, if "
+          + "finalization has already been started and did not finish yet."
+  )
+  private boolean force;
+
+  @Override
+  public void execute(ScmClient scmClient)
+      throws IOException, ExecutionException {
+    String upgradeClientID = "Upgrade-Client-" + UUID.randomUUID().toString();
+    try {
+      StatusAndMessages finalizationResponse =
+          scmClient.finalizeScmUpgrade(upgradeClientID);
+      if (isFinalized(finalizationResponse.status())){
+        System.out.println("Upgrade has already been finalized.");
+        emitExitMsg();
+        return;
+      } else if (!isStarting(finalizationResponse.status())){
+        System.err.println("Invalid response from Storage Container Manager.");
+        System.err.println(
+            "Current finalization status is: " + finalizationResponse.status()
+        );
+        throw new IOException("Exiting...");
+      }
+    } catch (UpgradeException e) {
+      handleInvalidRequestAfterInitiatingFinalization(force, e);
+    }
+    monitorAndWaitFinalization(scmClient, upgradeClientID);
+    return;
+  }
+
+  private void monitorAndWaitFinalization(ScmClient client,
+                                          String upgradeClientID) throws
+      ExecutionException {
+    ExecutorService exec = Executors.newSingleThreadExecutor();
+    Future<?> monitor =
+        exec.submit(new UpgradeMonitor(client, upgradeClientID, force));
+    try {
+      monitor.get();
+      emitFinishedMsg("Storage Container Manager");
+    } catch (CancellationException |InterruptedException e) {
+      emitCancellationMsg("Storage Container Manager");
+    } catch (ExecutionException e) {
+      emitGeneralErrorMsg();
+      throw e;
```

Comment:
I think, here we should just throw the cause of the ExecutionException wrapped into an IOException here.
I believe it is a better approach, as with that other subcommands are not forced to declare to throw two types of exceptions.

However as an alternative I am absolutely fine with declaring execute to throw an Exception, and do not mark specific Exceptions to be thrown.

Reply:
yup, I will make the change.